source/storage.ts

```diff
@@ -0,0 +1,34 @@
+export async function getTabsWaitingForInjection(): Promise<number[]> {
+	const fullStorage = await chrome.storage.session.getKeys();
+	return fullStorage
+		.filter(key => key.startsWith(storageKeyRoot))
+		.map(key => Number(key.slice(storageKeyRoot.length)));
+}
+
+export const storageKeyRoot = 'webext-inject-on-install:';
+export function getTabStorageKey(tabId: number) {
+	return `${storageKeyRoot}${tabId}`;
+}
+
+export async function isTabWaitingForInjection(tabId: number) {
+	if (ignoredTabs.has(tabId)) {
+		return false;
+	}
+
+	const key = getTabStorageKey(tabId);
+	const storage = await chrome.storage.session.get(key);
+	return key in storage;
+}
+
+// Local synchronous optimization to avoid multiple requests for the same tab
+export const ignoredTabs = new Set();
```

Comment:
This acts as "memoize while pending" and "module-level variable to ensure the wasn't a navigate". It covers:

```ts
// Repeated activations
void onActivated({tabId: 1})
void onActivated({tabId: 1})
```


```ts
// Navigate after activation
void onActivated({tabId: 1})
void onCommitted({tabId: 1})
```

Note that both are again extremely unlikely and do not affect PB due to its duplicate injection protection.

<img width="519" alt="Screenshot" src="https://github.com/user-attachments/assets/aba26150-e2ec-4436-9bc3-2bf3985d369a" />

It's way more likely that non-ready tabs are messaged.